File: simple_linear_regression/simple_linear_regression.R
library(ggplot2)
library(caret)
library(dplyr)

# Read data
salaryData <- read.csv('../datasets/Salary_Data.csv')

X <- salaryData[, 1]
y <- salaryData[, 2]

set.seed(123)
# Calculate random indexes and split into two groups. Train data will include 70% of data
train_indexes <- createDataPartition(y, times = 1, p = 0.7, list = FALSE)
X_train <- X[train_indexes]
X_test <- X[-train_indexes]
y_train <- y[train_indexes]
y_test <- y[-train_indexes]

# Train the model
model <- lm(y_train ~ X_train, salaryData)
print(coefficients(model))

# Test the model
y_pred <- predict(model, newdata = data.frame(X_train = X_test))

# Get model accuracy
mse <- mean((y_test - y_pred)^2)
print(paste("Mean Squared Error:", mse))
r2 <- as.numeric(cor(y_test, y_pred)^2)
print(paste("R-squared:", r2))

# Draw it
plot(y_train ~ X_train, data = salaryData)
abline(model)

ggplot() +
  geom_point(data = data.frame(X_test, y_test), aes(x = X_test, y = y_test), color = "lightblue") +
  geom_line(data = data.frame(X_test, y_test), aes(x = X_test, y = y_test), color = "blue") +
  geom_point(data = data.frame(X_test, y_pred), aes(x = X_test, y = y_pred), color = "lightgreen") +
  geom_line(data = data.frame(X_test, y_pred), aes(x = X_test, y = y_pred), color = "green") +
  labs(title = "Salary VS Experience", x = "Experience (years)", y = "Salary")

File: simple_linear_regression/simple_linear_regression.py
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Read data
dataset = pd.read_csv('../datasets/Salary_Data.csv')

X = dataset.iloc[:,:-1].values
y = dataset.iloc[:,1].values

# Calculate random indexes and split into two groups. Train data will include 70% of data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.7, random_state=0)

# Train the model
model = LinearRegression()
model.fit(X_train, y_train)

# Test the model
y_pred = model.predict(X_test)

# Get model accuracy
mse = mean_squared_error(y_test, y_pred)
print("Mean Squared Error:", mse)
r2 = r2_score(y_test, y_pred)
print("R-squared:", r2)

# Draw it
plt.scatter(X_train, y_train, color='red')
plt.plot(X_train, model.predict(X_train),color='blue')
plt.title('Salary VS Experience (Training data)')
plt.xlabel('Experience (years)')
plt.ylabel('Salary')
plt.show()
plt.scatter(X_test, y_test, color='red')
plt.plot(X_test, model.predict(X_test),color='blue')
plt.title('Salary VS Experience (Test data)')
plt.xlabel('Experience (years)')
plt.ylabel('Salary')
plt.show()